app/user.go

```diff
@@ -2495,7 +2496,32 @@ func (a *App) UpdateThreadReadForUserByPost(currentSessionId, userID, teamID, th
 		return nil, model.NewAppError("UpdateThreadReadForUser", "app.user.update_thread_read_for_user_by_post.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	return a.UpdateThreadReadForUser(currentSessionId, userID, teamID, threadID, post.CreateAt-1)
+	lastViewed := post.CreateAt - 1
+	if post.UserId == userID {
+		posts, nErr := a.Srv().Store.Thread().GetPosts(threadID, post.CreateAt)
+		if nErr != nil {
+			return nil, model.NewAppError("UpdateThreadReadForUser", "app.user.update_thread_read_for_user_by_post.app_error", nil, nErr.Error(), http.StatusInternalServerError)
+		}
+
+		sort.Slice(posts, func(i, j int) bool {
+			return posts[i].CreateAt < posts[j].CreateAt
+		})
+
+		found := false
+		for _, threadPost := range posts {
+			if threadPost.UserId != userID {
+				lastViewed = threadPost.CreateAt - 1
+				found = true
+				break
+			}
+		}
+
+		if !found {
+			lastViewed = posts[len(posts)-1].CreateAt + 1
+		}
+	}
```

Comment:
I think it would be better to do this in SQL (the store) instead of getting posts from store, sorting them, and then looping over them in the app layer.

Reply:
sounds good! I will update the PR.